### PR TITLE
MeterTags wrapper to support multiple MeterTag-s on a single method parameter

### LINKED ISSUE #4081: Improve MeterTag to allow extracting multiple keys/value tags
**Please describe the feature request.**
The '@MeterTag' feature lets you add dynamic tags, taking a value from a parameter. Currently, you can only extract one key/value pair because annotation can't be repeated and I did not see a way to define a list of key/value pairs. 
It would be helpful to allow for extracting multiple key-value pairs.

**Rationale**
For APIs with numerous fields, it would be beneficial to create multiple tags for improved filtering.

**Additional context**

### PR DESCRIPTION
**Pull request by smaxx**
…ethod parameter

Resolves #4081

### CODE REVIEW

**Review by marcingrzejszczak — APPROVED** [review]

File: micrometer-core/src/main/java/io/micrometer/core/aop/MeterTags.java
Comment (jonatan-ivanov) [inline micrometer-core/src/main/java/io/micrometer/core/aop/MeterTags.java:36]: We usually don't add new features after our RC release.
I think this should go to `1.14.0` instead of `1.13.0` (in less than a week).

File: micrometer-commons/src/main/java/io/micrometer/common/annotation/AnnotationUtils.java
Comment (jonatan-ivanov) [inline micrometer-commons/src/main/java/io/micrometer/common/annotation/AnnotationUtils.java:43]: I appreciate the refactoring effort especially since the previous version bothers me a little:
- `Annotation[][] parameters` should have been `parameterAnnotations`
- `Annotation[] parameter` should have been `annotations`
- `Annotation parameter2` should have been `annotation`
- incrementing the index separately

This refactor fixes these concerns but it also introduces a few more.
I don't think that doing this is a good usage of the Stream API (it might be also less performant), what do you think about using a for loop with an index instead of an `IntStream`?
Something like this I guess:
```java
Parameter[] parameters = method.getParameters();
for (int i = 0; i < parameters.length; i++) {
    Parameter parameter = parameters[i];
    Annotation[] annotations = parameter.getAnnotationsByType(annotationClazz);
    if (annotations.length == 0) {
        return Collections.emptyList();
    }
    else {
        return Arrays.stream(annotations)
            .map(annotation -> new AnnotatedParameter(i, annotation, args[i]))
            .collect(Collectors.toList());
    }
}
```

File: micrometer-core/src/main/java/io/micrometer/core/aop/MeterTags.java
Comment (jonatan-ivanov) [inline micrometer-core/src/main/java/io/micrometer/core/aop/MeterTags.java:28]: You can remove `<p>` I think.

File: micrometer-core/src/test/java/io/micrometer/core/aop/TimedAspectTest.java
Comment (jonatan-ivanov) [inline micrometer-core/src/test/java/io/micrometer/core/aop/TimedAspectTest.java:550]: Do you think it would worth adding one more test case that uses `MeterTags` explicitly (instead of Java 8's repeatable annotations)?

**Review by jonatan-ivanov — CHANGES_REQUESTED** [review]

File: micrometer-core/src/test/java/io/micrometer/core/aop/TimedAspectTest.java
Comment (smaxx) [reply micrometer-core/src/test/java/io/micrometer/core/aop/TimedAspectTest.java]: good point, added

File: micrometer-commons/src/main/java/io/micrometer/common/annotation/AnnotationUtils.java
Comment (smaxx) [reply micrometer-commons/src/main/java/io/micrometer/common/annotation/AnnotationUtils.java:41]: well, honestly i would argue about the approach :) Stream API-based approach could look like an overhead however it is always a matter of doing performance analysis first and then doing the conclusions especially now when we have corresponding tools like JMH. And unfortunately here we don't use extra libraries which would hide the additional complexity like StreamEx(`EntryStream.of(list)` will give you a stream of tuples where list element is associated with its index, or Guava's zipWithIndex). So this approach with IntStream doesn't seem to be breaking the main "rules" if such exist with regrads to how and where to use Stream API and in addition to that i remember i saw benchmark results where it was stated that on modern JVMs performance of Stream API is really close to `for` loops.

Anyway i reworked this place a bit to the suggested approach. Please see the result. Thank you so much for your comments and feedback

**Review by jonatan-ivanov — APPROVED** [review]

**Comment by jonatan-ivanov** [comment]
We can merge this next week after we released 1.13.0.

**Comment by shakuzen** [comment]
This is failing on `main` after merging. I'm looking into it now.

**Comment by shakuzen** [comment]
Looks like [this commit](https://github.com/micrometer-metrics/micrometer/commit/2069e9292e3649b98df77d5bfc7f58f475c78e07) wasn't in the builds run for this PR and that is making the difference in the tests failing. Though I think it's actually the implementation of `mergeAnnotatedParameters` in `AnnotationHandler` that should be fixed. In the failing tests, it is resulting in only one of the `MeterTag` annotations (the first one) being used.

**Comment by shakuzen** [comment]
I've reverted this change in 1f208955fa941e5648cbbd71fe8c33d965c155fb on `main`. We need to get it working with the latest code on main as pointed out in my previous comment. I'd reopen this pull request if I had the option. @smaxx would you mind opening a new pull request for this and we can figure out how to get everything working?

**Comment by smaxx** [comment]
> I've reverted this change in [1f20895](https://github.com/micrometer-metrics/micrometer/commit/1f208955fa941e5648cbbd71fe8c33d965c155fb) on `main`. We need to get it working with the latest code on main as pointed out in my previous comment. I'd reopen this pull request if I had the option. @smaxx would you mind opening a new pull request for this and we can figure out how to get everything working?

sure, will integrate the latest `main` and open a new PR today

**Pull request merged**
